src/engine/sparqlExpressions/StdevExpression.h

```diff
@@ -0,0 +1,163 @@
+#include "engine/sparqlExpressions/NaryExpression.h"
+#include "engine/sparqlExpressions/SparqlExpression.h"
+#include "engine/sparqlExpressions/SparqlExpressionTypes.h"
+#include "engine/sparqlExpressions/SparqlExpressionValueGetters.h"
+#include "global/ValueId.h"
+
+namespace sparqlExpression {
+
+namespace detail {
+
+/// The STDEV Expression
+
+// Helper function to extract a double from a NumericValue variant
+auto inline numValToDouble =
+    []<typename T>(T&& value) -> std::optional<double> {
+  if constexpr (ad_utility::isSimilar<T, double> ||
+                ad_utility::isSimilar<T, int64_t>) {
+    return static_cast<double>(value);
+  } else {
+    return std::nullopt;
+  }
+};
+
+// Helper expression: The individual deviation squares. A DeviationExpression
+// over X corresponds to the value (X - AVG(X))^2.
+class DeviationExpression : public SparqlExpression {
+ private:
+  Ptr child_;
+  bool distinct_;
+
+ public:
+  DeviationExpression(bool distinct, Ptr&& child)
+      : child_{std::move(child)}, distinct_{distinct} {}
+
+  // __________________________________________________________________________
+  ExpressionResult evaluate(EvaluationContext* context) const override {
+    auto impl =
+        [this, context](SingleExpressionResult auto&& el) -> ExpressionResult {
+      // Prepare space for result
+      VectorWithMemoryLimit<IdOrLiteralOrIri> exprResult{context->_allocator};
+      std::fill_n(std::back_inserter(exprResult), context->size(),
+                  IdOrLiteralOrIri{Id::makeUndefined()});
+
+      auto devImpl = [&exprResult, context](auto generator) {
+        double sum = 0.0;
+        std::vector<double> childResults = {};
+
+        // Collect values as doubles
+        for (auto& inp : generator) {
+          const auto& s = detail::NumericValueGetter{}(std::move(inp), context);
+          auto v = std::visit(numValToDouble, s);
+          if (v.has_value()) {
+            childResults.push_back(v.value());
+            sum += v.value();
+          }
+          context->cancellationHandle_->throwIfCancelled();
+        }
+
+        // Calculate squared deviation and save for result
+        double avg = sum / static_cast<double>(context->size());
+        for (size_t i = 0; i < childResults.size(); i++) {
+          exprResult.at(i) = IdOrLiteralOrIri{
+              ValueId::makeFromDouble(std::pow(childResults.at(i) - avg, 2))};
+        }
+      };
+
+      auto generator =
+          detail::makeGenerator(AD_FWD(el), context->size(), context);
+      if (distinct_) {
+        context->cancellationHandle_->throwIfCancelled();
+        devImpl(detail::getUniqueElements(context, context->size(),
+                                          std::move(generator)));
+      } else {
+        devImpl(std::move(generator));
+      }
+
+      return exprResult;
+    };
+    auto childRes = child_->evaluate(context);
+    return std::visit(impl, std::move(childRes));
```

Comment:
It's a shame that we currently cannot simply reuse the SUM and AVG expressions for this without copying the result of X, but this  code is not too bad.